pkg/gadgets/helpers.go

```diff
@@ -127,6 +129,17 @@ func IPStringFromBytes(ipBytes [16]byte, ipType int) string {
 	}
 }
 
+// IPStringToUint32 converts an IP address (IPv4 only) string to a uint32 in
+// little endian format.
+func IPStringToUint32(ipAddr string) (uint32, error) {
```

Comment:
Mmmm I made some tests during the development, and I found that I needed the values for filtering in little-endian to match:

```bash
$ sudo ./ig profile tcprtt --byladdr --laddr 127.0.0.1
$ sudo bpftool map dump name hists | jq '.[0].key'
16777343
$ sudo bpftool map dump name .rodata | jq '.[0].value.".rodata"[]|select (."targ_saddr")'
{
  "targ_saddr": 16777343
}
$ printf 'obase=16;16777343'|bc -qi
obase=16;16777343
100007F
$ printf 'ibase=16;7F'|bc -qi
ibase=16;7F
127
```

The value reported as the map key is in little-endian:

https://github.com/inspektor-gadget/inspektor-gadget/blob/afc30ea4671c227d4b6871cbc1de43f1216a4ecd/pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c#L49-L50

And it is the same used for the filtering:

https://github.com/inspektor-gadget/inspektor-gadget/blob/afc30ea4671c227d4b6871cbc1de43f1216a4ecd/pkg/gadgets/profile/tcprtt/tracer/bpf/tcprtt.bpf.c#L40-L41

At this point, if the IP and port are supposed to be in big-endian (network byte order), the only possibility is that `BPF_CORE_READ` is doing the conversion. WDYT?

Reply:
I don't think that BPF_CORE_READ is doing the conversion. But I don't have other explanations...

Reply:
I still don't understand how is it possible that `binary.LittleEndian.Uint32(net.ParseIP("127.0.0.1").To4())` gives me 0x100007f, which according to your explanation, is in big-endian?

Reply:
I suppose it is because `.To4()` stores the bytes already in little-endian, so `binary.LittleEndian.Uint32` is actually transforming it to big-endian. Right?

Reply:
I mean, this program doesn't make sense to me:

```golang
package main

import (
	"encoding/binary"
	"fmt"
	"net"
)

func main() {
	ip := net.ParseIP("127.0.0.1").To4()
	fmt.Printf("%X\n", binary.LittleEndian.Uint32(ip)) # Prints big-endian representation: 100007F
	fmt.Printf("%X\n", binary.BigEndian.Uint32(ip)) # Prints little-endian representation: 7F000001
}
```

Reply:
7F000001 is Big-Endian, 0100007F is Little-Endian, so it looks right

Reply:
Mmmm we concluded that the big-endian representation of "127.0.0.1" is 100007F 😅

Reply:
The `bpf_printk` might reverse it as well when used with `%x`.

To show the underlying memory in Go:
```go
	ip := net.ParseIP("127.0.0.1").To4()
	fmt.Printf("%X\n", []byte(ip[:])) // 7F 00 00 01 // BigEndian in-memory

	ip2be := binary.BigEndian.Uint32(ip) // Convert from BE to native (LE)
	fmt.Printf("%X\n", ip2be) // 7F 00 00 01 // %X takes native (LE) into account and reverses it
	fmt.Printf("%X\n", *(*[4]byte)(unsafe.Pointer(&ip2be))) // 01 00 00 7F // but the underlying memory is reversed
```

Reply:
@flyth right, `bpf_printk("%0x".. )` reverses it. Its indeed `{0x7F, 0x00, 0x00, 0x01}` in memory (big endian)

It seems the whole confusion is we're using `sudo bpftool map dump name .rodata|jq '.[0].value.".rodata"[]|select (."targ_saddr")'` to print an integer stored in big endian on a little endian machine, so it prints it "reversed". For the `127.0.0.1` IP it prints `16777343`, that is `0x100007f`. 

Let's look at the content of the map in memory. Constants are packed by the compiler in a DATASEC, so we can look at this  layout using BTF:

```
$ sudo bpftool map show name .rodata 
104: array  name .rodata  flags 0x80
	key 4B  value 85B  max_entries 1  memlock 4096B
	btf_id 370  frozen

# btf_id 370 is all what we need 

$ sudo bpftool btf dump id 370
[1] CONST '(anon)' type_id=7
[2] CONST '(anon)' type_id=10
[3] CONST '(anon)' type_id=13
[4] DATASEC '.rodata' size=85 vlen=11
	type_id=29 offset=0 size=1 (VAR 'targ_laddr_hist')
	type_id=28 offset=1 size=1 (VAR 'targ_raddr_hist')
	type_id=27 offset=2 size=2 (VAR 'targ_sport')
	type_id=26 offset=4 size=2 (VAR 'targ_dport')
	type_id=25 offset=8 size=4 (VAR 'targ_saddr')
	type_id=24 offset=12 size=4 (VAR 'targ_daddr')
	type_id=23 offset=16 size=1 (VAR 'targ_ms')
	type_id=22 offset=17 size=15 (VAR 'handle_tcp_rcv_established.____fmt')
	type_id=21 offset=32 size=19 (VAR 'handle_tcp_rcv_established.____fmt.1')
	type_id=20 offset=51 size=15 (VAR 'handle_tcp_rcv_established.____fmt.2')
	type_id=18 offset=66 size=19 (VAR 'handle_tcp_rcv_established.____fmt.3')
...
```

`targ_saddr` starts in byte 8 and has a size of 4, let's dump the map memory.

```bash
$ sudo bpftool map dump name .rodata -j 
[{"key":["0x00","0x00","0x00","0x00"],"value":["0x01","0x00","0x00","0x00","0x00","0x00","0x00","0x00","0x7f","0x00","0x00","0x01","0x00","0x00","0x00","0x00","0x00","0x69","0x6e","0x65","0x74","0x5f","0x73","0x61","0x64","0x64","0x72","0x20","0x25","0x30","0x78","0x00","0x69","0x6e","0x65","0x74","0x5f","0x73","0x61","0x64","0x64","0x72","0x5b","0x25","0x64","0x5d","0x20","0x25","0x30","0x78","0x00","0x74","0x61","0x72","0x67","0x5f","0x73","0x61","0x64","0x64","0x72","0x20","0x25","0x30","0x78","0x00","0x74","0x61","0x72","0x67","0x5f","0x73","0x61","0x64","0x64","0x72","0x5b","0x25","0x64","0x5d","0x20","0x25","0x30","0x78","0x00"],"formatted":{"value":{".rodata":[{"targ_laddr_hist":true},{"targ_raddr_hist":false},{"targ_sport":0},{"targ_dport":0},{"targ_saddr":16777343},{"targ_daddr":0},{"targ_ms":false},{"handle_tcp_rcv_established.____fmt":"inet_saddr %0x"},{"handle_tcp_rcv_established.____fmt.1":"inet_saddr[%d] %0x"},{"handle_tcp_rcv_established.____fmt.2":"targ_saddr %0x"},{"handle_tcp_rcv_established.____fmt.3":"targ_saddr[%d] %0x"}]}}}]
``` 

```
"0x01","0x00","0x00","0x00","0x00","0x00","0x00","0x00","0x7f","0x00","0x00","0x01","0x00"
  0   |  1   |  2   |  3   |  4   |  5   |  6   |  7   |  8   |  9   |  10  |  11  |  12 |
```

bytes 8 to 11 correspond to `targ_saddr`, so it's saved in memory as `{0x7f,0x00,0x00,0x01}`, i.e, it's big endian. 

So I think we can conclude:
- inet_saddr is stored in big endian notation (for sure it's since it's __be32 on kernel source code)
- `IPStringToUint32` produces a big endian representation. 
- `targ_saddr` is saved on big endian. (bpftool dump output is "misleading" on a little endian machine)

Did I get it right this time?

Reply:
> IPStringToUint32 produces a big endian representation.

Yeah, in our case (using binary.LittleEndian on intel/arm) it's usually a no-op, so it'll stay big endian under the hood - but it's not portable. On a big endian system, it would reverse the value. So I think it's better to cast - we know the slice is in big endian and we know we'll compare it against big endian values, the uint32 is just our means of transport. It would probably be even better to use char[4] and just copy the bytes from IP, but then comparison isn't as clean, right?

Reply:
>  (using binary.LittleEndian on intel/arm) it's usually a no-op, so it'll stay big endian under the hood - but it's not portable. On a big endian system, it would reverse the value

Are you sure?

It seems to me from the code that this function behaves the same regardless the architecture it runs in. -> https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/encoding/binary/binary.go;l=79

IIUC, `binary.LittleEndian` refers to the endianess of the type of pass in the byte array, and not the machine where it's running.

Reply:
AFAIU, the difference is in the uint32() used in that code and the shifting done - depending on the machine used, it will change the underlying memory (because under the hood, the shifting direction in terms of the memory is different). If you feed binary.LittleEndian a slice of memory that stores a uint32 on a LE machine, it won't change the order in memory (so it's basically just an expensive cast). On a BE machine, it will reverse the order, though (so not only a cast).

If you'd use the same way later on to convert again, you'd of course have the correct value again (because it's either again a no-op (LE) or will be reversed (BE)), but as we only forward the "converted" value to BPF (cilium/ebpf just forwards using the native endian, see [here](https://github.com/cilium/ebpf/blob/master/marshalers.go#L61)), this "fixing" would have to be done inside the BPF code. That's why I suggest just casting in the first place instead of converting. We know both sides (our IP().To4() and the __be32 from eBPF) share the same endianness, so we can just leave the memory untouched and not care how it is interpreted as integers.

Reply:
>>  (using binary.LittleEndian on intel/arm) it's usually a no-op, so it'll stay big endian under the hood - but it's not portable. On a big endian system, it would reverse the value
>>
> Are you sure?
>
For sure. In fact, that's the only way to explain what we have seen so far. Otherwise, how would we explain that `IPStringToUint32` converts to big-endian if `ip.To4()` returns a byte array in big-endian? Our answer now is: `binary.LittleEndian` does nothing in our little-endian machines.

Reply:
> Our answer now is: binary.LittleEndian does nothing in our little-endian machines.

My question was: does it behave differently on a big-endian machine? 

Anyway folks, I'm still too confused about it, I'll take an endianess 101 course and once I have a good understanding I'll comment. I'd say for the time being we can move forward if this works fine on arm64 and amd64 since those are the only platforms we support so far.